tools/clang/lib/SPIRV/SPIRVEmitter.cpp

```diff
@@ -3852,6 +3874,118 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
   return 0;
 }
 
+uint32_t
+SPIRVEmitter::processIntrinsicInterlockedMethod(const CallExpr *expr,
+                                                hlsl::IntrinsicOp opcode) {
+  // The signature of intrinsic atomic methods are:
+  // void Interlocked*(in R dest, in T value);
+  // void Interlocked*(in R dest, in T value, out T original_value);
+
+  // Note: ALL Interlocked*() methods are forced to have an unsigned integer
+  // 'value'. Meaning, T is forced to be 'unsigned int'. If the provided
+  // parameter is not an unsigned integer, DXC inserts an 'ImplicitCastExpr' to
+  // convert it to unsigned integer. OpAtomicIAdd (and other SPIR-V OpAtomic*
+  // instructions) require that the pointee in 'dest' to be of the same type as
+  // T. This will result in an invalid SPIR-V if 'dest' is a signed integer
+  // typed resource such as RWTexture1D<int>. For example, the following
+  // OpAtomicIAdd is invalid because the pointee type defined in %1 is a signed
+  // integer, while the value passed to atomic add (%3) is an unsigned integer.
+  //
+  //  %_ptr_Image_int = OpTypePointer Image %int
+  //  %1 = OpImageTexelPointer %_ptr_Image_int %RWTexture1D_int %index %uint_0
+  //  %2 = OpLoad %int %value
+  //  %3 = OpBitcast %uint %2   <-------- Inserted by DXC
+  //  %4 = OpAtomicIAdd %int %1 %uint_1 %uint_0 %3
+  //
+  // In such cases, we bypass the forced IntegralCast.
+  // Moreover, DXC does not add a cast AST node to cast uint to int where
+  // necessary. To ensure SPIR-V validity, we add that where necessary.
+
+  const uint32_t zero = theBuilder.getConstantUint32(0);
+  const uint32_t scope = theBuilder.getConstantUint32(1); // Device
+  const auto dest = expr->getArg(0);
```

Comment:
Nit: It's better to use `auto*` if the right side is a pointer. Although this doesn't really matter since `auto` is expanded to a pointer itself, it reads like we are doing a copy here.

Reply:
done